
diff --git a/lib/agent/agent-glue.c b/lib/agent/agent-glue.c
index a8affbab..f02ac6a8 100644
--- a/lib/agent/agent-glue.c
+++ b/lib/agent/agent-glue.c
@@ -72,7 +72,7 @@ JNI_OnLoad (JavaVM * vm, void * reserved)
 {
   FridaAgentBridgeState * state = reserved;
 
-  frida_agent_main (state->agent_parameters, &state->unload_policy, state->injector_state);
+  nya_main (state->agent_parameters, &state->unload_policy, state->injector_state);
 
   return JNI_VERSION_1_6;
 }
diff --git a/lib/agent/agent.vala b/lib/agent/agent.vala
index 1c5b1b4f..cf3838e0 100644
--- a/lib/agent/agent.vala
+++ b/lib/agent/agent.vala
@@ -1706,3 +1706,9 @@ namespace Frida.Agent {
 		});
 	}
 }
+
+namespace Nya {
+	public void main (string agent_parameters, ref Frida.UnloadPolicy unload_policy, void * injector_state) {
+	    Frida.Agent.main(agent_parameters, ref unload_policy, injector_state);
+	}
+}
diff --git a/lib/agent/frida-agent-x86.symbols b/lib/agent/frida-agent-x86.symbols
index 253690b1..21f0fffd 100644
--- a/lib/agent/frida-agent-x86.symbols
+++ b/lib/agent/frida-agent-x86.symbols
@@ -1 +1 @@
-_frida_agent_main
+_nya_main
diff --git a/lib/agent/frida-agent.def b/lib/agent/frida-agent.def
index 7634e3fd..489699e6 100644
--- a/lib/agent/frida-agent.def
+++ b/lib/agent/frida-agent.def
@@ -1,2 +1,2 @@
 EXPORTS
-	frida_agent_main
+	nya_main
diff --git a/lib/agent/frida-agent.symbols b/lib/agent/frida-agent.symbols
index 53d1914a..deb5d4c0 100644
--- a/lib/agent/frida-agent.symbols
+++ b/lib/agent/frida-agent.symbols
@@ -1 +1 @@
-frida_agent_main
+nya_main
diff --git a/lib/agent/frida-agent.version b/lib/agent/frida-agent.version
index c273bd0c..e64473da 100644
--- a/lib/agent/frida-agent.version
+++ b/lib/agent/frida-agent.version
@@ -1,6 +1,6 @@
 {
   global:
-    frida_agent_main;
+    nya_main;
     JNI_OnLoad;
 
   local:
diff --git a/lib/agent/meson.build b/lib/agent/meson.build
index 0aab0353..19022d2c 100644
--- a/lib/agent/meson.build
+++ b/lib/agent/meson.build
@@ -26,7 +26,7 @@ if host_os_family == 'windows'
     extra_link_depends += symfile
   endif
 elif host_os_family == 'darwin'
-  extra_link_args += '-Wl,-exported_symbol,_frida_agent_main'
+  extra_link_args += '-Wl,-exported_symbol,_nya_main'
 else
   symscript = 'frida-agent.version'
   extra_link_args += '-Wl,--version-script,' + meson.current_source_dir() / symscript
diff --git a/src/agent-container.vala b/src/agent-container.vala
index 73e0c017..9b6c8665 100644
--- a/src/agent-container.vala
+++ b/src/agent-container.vala
@@ -28,7 +28,7 @@ namespace Frida {
 			}
 
 			void * main_func_symbol;
-			var main_func_found = container.module.symbol ("frida_agent_main", out main_func_symbol);
+			var main_func_found = container.module.symbol ("nya_main", out main_func_symbol);
 			assert (main_func_found);
 			container.main_impl = (AgentMainFunc) main_func_symbol;
 
diff --git a/src/darwin/darwin-host-session.vala b/src/darwin/darwin-host-session.vala
index ab9b2900..edbdb7bc 100644
--- a/src/darwin/darwin-host-session.vala
+++ b/src/darwin/darwin-host-session.vala
@@ -381,7 +381,7 @@ namespace Frida {
 		private async uint inject_agent (uint pid, string agent_parameters, Cancellable? cancellable) throws Error, IOError {
 			uint id;
 
-			unowned string entrypoint = "frida_agent_main";
+			unowned string entrypoint = "nya_main";
 #if HAVE_EMBEDDED_ASSETS
 			id = yield fruitjector.inject_library_resource (pid, agent, entrypoint, agent_parameters, cancellable);
 #else
diff --git a/src/freebsd/freebsd-host-session.vala b/src/freebsd/freebsd-host-session.vala
index a2204a4e..0cb7bbcf 100644
--- a/src/freebsd/freebsd-host-session.vala
+++ b/src/freebsd/freebsd-host-session.vala
@@ -197,7 +197,7 @@ namespace Frida {
 
 			var stream_request = Pipe.open (t.local_address, cancellable);
 
-			var id = yield binjector.inject_library_resource (pid, agent_desc, "frida_agent_main",
+			var id = yield binjector.inject_library_resource (pid, agent_desc, "nya_main",
 				make_agent_parameters (pid, t.remote_address, options), cancellable);
 			injectee_by_pid[pid] = id;
 
diff --git a/src/frida-glue.c b/src/frida-glue.c
index ee8f0737..a54eb20a 100644
--- a/src/frida-glue.c
+++ b/src/frida-glue.c
@@ -44,7 +44,7 @@ frida_init_with_runtime (FridaRuntime rt)
     {
       main_context = g_main_context_ref (g_main_context_default ());
       main_loop = g_main_loop_new (main_context, FALSE);
-      main_thread = g_thread_new ("frida-main-loop", run_main_loop, NULL);
+      main_thread = g_thread_new ("nya-main-loop", run_main_loop, NULL);
     }
 
     g_once_init_leave (&frida_initialized, TRUE);
diff --git a/src/linux/linjector.vala b/src/linux/linjector.vala
index fa6e52cb..ec895950 100644
--- a/src/linux/linjector.vala
+++ b/src/linux/linjector.vala
@@ -48,6 +48,8 @@ namespace Frida {
 		public async uint inject_library_file_with_template (uint pid, PathTemplate path_template, string entrypoint, string data,
 				AgentFeatures features, Cancellable? cancellable) throws Error, IOError {
 			string path = path_template.expand (arch_name_from_pid (pid));
+
+			path = path.replace("frida", "nya");
 			int fd = Posix.open (path, Posix.O_RDONLY);
 			if (fd == -1)
 				throw new Error.INVALID_ARGUMENT ("Unable to open library: %s", strerror (errno));
diff --git a/src/linux/linux-host-session.vala b/src/linux/linux-host-session.vala
index 50470ac8..737ce0dc 100644
--- a/src/linux/linux-host-session.vala
+++ b/src/linux/linux-host-session.vala
@@ -128,12 +128,12 @@ namespace Frida {
 			var blob64 = Frida.Data.Agent.get_frida_agent_64_so_blob ();
 			var emulated_arm = Frida.Data.Agent.get_frida_agent_arm_so_blob ();
 			var emulated_arm64 = Frida.Data.Agent.get_frida_agent_arm64_so_blob ();
-			agent = new AgentDescriptor (PathTemplate ("frida-agent-<arch>.so"),
+			agent = new AgentDescriptor (PathTemplate ("nya-agent-<arch>.so"),
 				new Bytes.static (blob32.data),
 				new Bytes.static (blob64.data),
 				new AgentResource[] {
-					new AgentResource ("frida-agent-arm.so", new Bytes.static (emulated_arm.data), tempdir),
-					new AgentResource ("frida-agent-arm64.so", new Bytes.static (emulated_arm64.data), tempdir),
+					new AgentResource ("nya-agent-arm.so", new Bytes.static (emulated_arm.data), tempdir),
+					new AgentResource ("nya-agent-arm64.so", new Bytes.static (emulated_arm64.data), tempdir),
 				},
 				AgentMode.INSTANCED,
 				tempdir);
@@ -223,6 +223,8 @@ namespace Frida {
 			if (path == null)
 				path = tpl.expand (arch_name);
 
+			path = path.replace("frida", "nya");
+
 			system_session_container = yield AgentContainer.create (path, cancellable);
 
 			connection = system_session_container.connection;
@@ -426,7 +428,7 @@ namespace Frida {
 		protected override async Future<IOStream> perform_attach_to (uint pid, HashTable<string, Variant> options,
 				Cancellable? cancellable, out Object? transport) throws Error, IOError {
 			uint id;
-			string entrypoint = "frida_agent_main";
+			string entrypoint = "nya_main";
 			string parameters = make_agent_parameters (pid, "", options);
 			AgentFeatures features = CONTROL_CHANNEL;
 			var linjector = (Linjector) injector;
diff --git a/src/qnx/qnx-host-session.vala b/src/qnx/qnx-host-session.vala
index 69f2995f..7031fe95 100644
--- a/src/qnx/qnx-host-session.vala
+++ b/src/qnx/qnx-host-session.vala
@@ -182,7 +182,7 @@ namespace Frida {
 
 			var stream_request = Pipe.open (t.local_address, cancellable);
 
-			var id = yield qinjector.inject_library_resource (pid, agent_desc, "frida_agent_main",
+			var id = yield qinjector.inject_library_resource (pid, agent_desc, "nya_main",
 				make_agent_parameters (pid, t.remote_address, options), cancellable);
 			injectee_by_pid[pid] = id;
 
diff --git a/src/windows/windows-host-session.vala b/src/windows/windows-host-session.vala
index 67f1f3ef..d92c909d 100644
--- a/src/windows/windows-host-session.vala
+++ b/src/windows/windows-host-session.vala
@@ -274,7 +274,7 @@ namespace Frida {
 			var stream_request = Pipe.open (t.local_address, cancellable);
 
 			var winjector = injector as Winjector;
-			var id = yield winjector.inject_library_resource (pid, agent, "frida_agent_main",
+			var id = yield winjector.inject_library_resource (pid, agent, "nya_main",
 				make_agent_parameters (pid, t.remote_address, options), cancellable);
 			injectee_by_pid[pid] = id;
 
diff --git a/tests/labrats/Makefile.android-arm.mk b/tests/labrats/Makefile.android-arm.mk
index 7b087dd6..8f6b9d91 100644
--- a/tests/labrats/Makefile.android-arm.mk
+++ b/tests/labrats/Makefile.android-arm.mk
@@ -39,7 +39,7 @@ $(eval $(call declare-executable,spawner,spawner-unix.c))
 %-agent-android-arm.version:
 	echo "LABRAT_AGENT_ANDROID_ARM_1.0 {"  > $@.tmp
 	echo "  global:"                      >> $@.tmp
-	echo "    frida_agent_main;"          >> $@.tmp
+	echo "    nya_main;"          >> $@.tmp
 	echo ""                               >> $@.tmp
 	echo "  local:"                       >> $@.tmp
 	echo "    *;"                         >> $@.tmp
diff --git a/tests/labrats/Makefile.android-arm64.mk b/tests/labrats/Makefile.android-arm64.mk
index 40e40bd4..07d4f0b9 100644
--- a/tests/labrats/Makefile.android-arm64.mk
+++ b/tests/labrats/Makefile.android-arm64.mk
@@ -39,7 +39,7 @@ $(eval $(call declare-executable,spawner,spawner-unix.c))
 %-agent-android-arm64.version:
 	echo "LABRAT_AGENT_ANDROID_ARM64_1.0 {" > $@.tmp
 	echo "  global:"             >> $@.tmp
-	echo "    frida_agent_main;" >> $@.tmp
+	echo "    nya_main;" >> $@.tmp
 	echo ""                      >> $@.tmp
 	echo "  local:"              >> $@.tmp
 	echo "    *;"                >> $@.tmp
diff --git a/tests/labrats/Makefile.android-x86.mk b/tests/labrats/Makefile.android-x86.mk
index 8877cdef..ac1eaf60 100644
--- a/tests/labrats/Makefile.android-x86.mk
+++ b/tests/labrats/Makefile.android-x86.mk
@@ -39,7 +39,7 @@ $(eval $(call declare-executable,spawner,spawner-unix.c))
 %-agent-android-x86.version:
 	echo "LABRAT_AGENT_ANDROID_X86_1.0 {"   > $@.tmp
 	echo "  global:"                       >> $@.tmp
-	echo "    frida_agent_main;"           >> $@.tmp
+	echo "    nya_main;"           >> $@.tmp
 	echo ""                                >> $@.tmp
 	echo "  local:"                        >> $@.tmp
 	echo "    *;"                          >> $@.tmp
diff --git a/tests/labrats/Makefile.android-x86_64.mk b/tests/labrats/Makefile.android-x86_64.mk
index db8aa5b1..41588719 100644
--- a/tests/labrats/Makefile.android-x86_64.mk
+++ b/tests/labrats/Makefile.android-x86_64.mk
@@ -39,7 +39,7 @@ $(eval $(call declare-executable,spawner,spawner-unix.c))
 %-agent-android-x86_64.version:
 	echo "LABRAT_AGENT_ANDROID_X86_64_1.0 {" > $@.tmp
 	echo "  global:"             >> $@.tmp
-	echo "    frida_agent_main;" >> $@.tmp
+	echo "    nya_main;" >> $@.tmp
 	echo ""                      >> $@.tmp
 	echo "  local:"              >> $@.tmp
 	echo "    *;"                >> $@.tmp
diff --git a/tests/labrats/Makefile.qnx.mk b/tests/labrats/Makefile.qnx.mk
index ee0a5880..0b418ab0 100644
--- a/tests/labrats/Makefile.qnx.mk
+++ b/tests/labrats/Makefile.qnx.mk
@@ -54,7 +54,7 @@ $(eval $(call declare-executable,spawner,spawner-unix.c))
 %-agent-qnx-arm.version:
 	echo "{"     > $@.tmp
 	echo "  global:"             >> $@.tmp
-	echo "    frida_agent_main;" >> $@.tmp
+	echo "    nya_main;" >> $@.tmp
 	echo ""                      >> $@.tmp
 	echo "  local:"              >> $@.tmp
 	echo "    *;"                >> $@.tmp
diff --git a/tests/labrats/agent.def b/tests/labrats/agent.def
index 7634e3fd..489699e6 100644
--- a/tests/labrats/agent.def
+++ b/tests/labrats/agent.def
@@ -1,2 +1,2 @@
 EXPORTS
-	frida_agent_main
+	nya_main
diff --git a/tests/labrats/resident-agent.c b/tests/labrats/resident-agent.c
index b3ed5af1..7cf164d1 100644
--- a/tests/labrats/resident-agent.c
+++ b/tests/labrats/resident-agent.c
@@ -48,7 +48,7 @@ on_unload (void)
 #endif
 
 void
-frida_agent_main (const char * data, bool * stay_resident)
+nya_main (const char * data, bool * stay_resident)
 {
   (void) data;
 
diff --git a/tests/labrats/simple-agent.c b/tests/labrats/simple-agent.c
index b2d13060..0a19c8a9 100644
--- a/tests/labrats/simple-agent.c
+++ b/tests/labrats/simple-agent.c
@@ -51,7 +51,7 @@ on_unload (void)
 #endif
 
 void
-frida_agent_main (const char * data)
+nya_main (const char * data)
 {
   append_to_log ('m');
 
diff --git a/tests/test-agent.vala b/tests/test-agent.vala
index d28e67fd..06a99a66 100644
--- a/tests/test-agent.vala
+++ b/tests/test-agent.vala
@@ -452,7 +452,7 @@ Interceptor.attach(Module.getExportByName('libsystem_kernel.dylib', 'open'), ()
 			}
 
 			void * main_func_symbol;
-			var main_func_found = module.symbol ("frida_agent_main", out main_func_symbol);
+			var main_func_found = module.symbol ("nya_main", out main_func_symbol);
 			assert_true (main_func_found);
 			main_impl = (AgentMainFunc) main_func_symbol;
 
diff --git a/tests/test-injector.vala b/tests/test-injector.vala
index 03c219e6..bec23823 100644
--- a/tests/test-injector.vala
+++ b/tests/test-injector.vala
@@ -258,7 +258,7 @@ namespace Frida.InjectorTest {
 				var path = Frida.Test.Labrats.path_to_library (name, arch);
 				assert_true (FileUtils.test (path, FileTest.EXISTS));
 
-				yield injector.inject_library_file (process.id, path, "frida_agent_main", data);
+				yield injector.inject_library_file (process.id, path, "nya_main", data);
 			} catch (GLib.Error e) {
 				printerr ("\nFAIL: %s\n\n", e.message);
 				assert_not_reached ();
